### PR TITLE
Live: Experimental HA with Redis

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

Grafana v8.1 will include experimental support for HA setup with Redis (added in #34851).

**Which issue(s) this PR fixes**:

Addresses a part of https://github.com/grafana/grafana/issues/35448#issuecomment-872224175

### CODE REVIEW

**Review by ryantxu — APPROVED** [review]
LGTM

File: docs/sources/administration/configuration.md [hunk 1]
```diff
@@ -1525,6 +1525,30 @@ For example:
 allowed_origins = "https://*.example.com"
 ```
 
+### ha_engine
+
+> **Note**: Available in Grafana v8.1 and later versions.
+
+**Experimental**
+
+The name of high-availability engine for Grafana Live. By default, it's not set. The only possible value is "redis".
```
Comment (achatterjee-grafana) [inline docs/sources/administration/configuration.md:1534]:
```suggestion
The high availability (HA) engine name for Grafana Live. By default, it's not set. The only possible value is "redis".
```

File: docs/sources/administration/configuration.md [hunk 2]
```diff
@@ -1525,6 +1525,30 @@ For example:
...
+
+More information in [Configure Grafana Live HA setup]({{< relref "../live/live-ha-setup.md" >}}).
```
Comment (achatterjee-grafana) [inline docs/sources/administration/configuration.md:1536]:
```suggestion
For more information, refer to [Configure Grafana Live HA setup]({{< relref "../live/live-ha-setup.md" >}}).
```

File: docs/sources/administration/configuration.md [hunk 3]
```diff
@@ -1525,6 +1525,30 @@ For example:
...
+
+### ha_engine_address
+
+> **Note**: Available in Grafana v8.1 and later versions.
+
+**Experimental**
+
+Address string of selected HA Live engine. For Redis it's a `host:port` string. Example:
```
Comment (achatterjee-grafana) [inline docs/sources/administration/configuration.md:1544]:
```suggestion
Address string of selected the high availability (HA) Live engine. For Redis, it's a `host:port` string. Example:
```

File: docs/sources/live/live-ha-setup.md [hunk 4]
```diff
@@ -7,10 +7,52 @@ weight = 130
 
 # Configure Grafana Live HA setup
 
-Live features in Grafana v8.0 are designed to work with a single Grafana server instance only. We will add the option for HA configuration in future Grafana releases to eliminate the current limitations.
+By default, Grafana Live uses in-memory data structures and in-memory PUB/SUB hub for handling subscriptions.
 
-Currently, if you have several Grafana server instances behind a load balancer, you may come across the following limitations:
+In HA Grafana setup which involves several Grafana server instances behind a load balancer you may come across the following limitations:
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:12]:
```suggestion
In a high availability Grafana setup involving several Grafana server instances behind a load balancer, you can find the following limitations:
```

File: docs/sources/live/live-ha-setup.md [hunk 5]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
 
 - Built-in features like dashboard change notifications will only be broadcasted to users connected to the same Grafana server process instance.
 - Streaming from Telegraf will deliver data only to clients connected to the same instance which received Telegraf data, active stream cache is not shared between different Grafana instances.
 - A separate unidirectional stream between Grafana and backend data source may be opened on different Grafana servers for the same channel.
+
+To eliminate these limitations Grafana v8.1 introduced an experimental Live HA engine which requires Redis to work.
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:18]:
```suggestion
To bypass these limitations, Grafana v8.1 has an experimental Live HA engine that requires Redis to work.
```

File: docs/sources/live/live-ha-setup.md [hunk 6]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
+
+## Configure Redis Live engine
+
+With Redis engine configured Grafana Live will keep its state in Redis and will use Redis PUB/SUB functionality to deliver messages to all subscribers throughout all Grafana server nodes.
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:22]:
```suggestion
When the Redis engine is configured, Grafana Live keeps its state in Redis and uses Redis PUB/SUB functionality to deliver messages to all subscribers throughout all Grafana server nodes.
```

File: docs/sources/live/live-ha-setup.md [hunk 7]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
+
+```
+[live]
+ha_engine = redis
+ha_engine_address = 127.0.0.1:6379
+```
+
+After running:
+
+- All built-in real-time notifications like dashboard changes will be delivered to all Grafana server instances and broadcasted to all subscribers.
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:34]:
```suggestion
- All built-in real-time notifications like dashboard changes are delivered to all Grafana server instances and broadcasted to all subscribers.
```

File: docs/sources/live/live-ha-setup.md [hunk 8]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
+- Streaming from Telegraf will deliver messages to all subscribers
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:35]:
```suggestion
- Streaming from Telegraf delivera messages to all subscribers.
```

File: docs/sources/live/live-ha-setup.md [hunk 9]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
+- A separate unidirectional stream between Grafana and backend data source will still be opened on different Grafana servers but publishing data to a channel will only deliver messages to instance subscribers thus publications from different instances on different machines won't result into duplicate data on panels.
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:36]:
```suggestion
- A separate unidirectional stream between Grafana and backend data source opens on different Grafana servers. Publishing data to a channel delivers messages to instance subscribers, as a result, publications from different instances on different machines do not produce duplicate data on panels.
```

File: docs/sources/live/live-ha-setup.md [hunk 10]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
+
+At the moment we only support single Redis node.
+
+> **Note:**  It's possible to use Redis Sentinel and Haproxy to achieve highly-available Redis setup. Redis nodes should be managed by [Redis Sentinel](https://redis.io/topics/sentinel) to achieve automatic failover. Haproxy configuration example:
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:40]:
```suggestion
> **Note:**  It's possible to use Redis Sentinel and Haproxy to achieve a highly available Redis setup. Redis nodes should be managed by [Redis Sentinel](https://redis.io/topics/sentinel) to achieve automatic failover. Haproxy configuration example:
```

File: docs/sources/live/live-ha-setup.md [hunk 11]
```diff
@@ -7,10 +7,52 @@ weight = 130
...
+>   option tcp-check
+>   tcp-check send PING\r\n
+>   tcp-check expect string +PONG
+>   tcp-check send info\ replication\r\n
+>   tcp-check expect string role:master
+>   tcp-check send QUIT\r\n
+>   tcp-check expect string +OK
+>   balance roundrobin
+> ```
+> Then point Grafana Live to Haproxy address:port.
```
Comment (achatterjee-grafana) [inline docs/sources/live/live-ha-setup.md:58]:
```suggestion
> Next, point Grafana Live to Haproxy address:port.
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added some review comments.

**Review by achatterjee-grafana — APPROVED** [review]
LGTM

**Pull request merged**